Web Inspector: support `console.profile` in `Worker`

dcrousso opened the pull request:
#### 98f9f65893366f159adec513808b596ab972a7dd
<pre>
Web Inspector: support `console.profile` in `Worker`
<a href="https://bugs.webkit.org/show_bug.cgi?id=287069">https://bugs.webkit.org/show_bug.cgi?id=287069</a>

Reviewed by BJ Burg.

* Source/WebCore/workers/WorkerConsoleClient.cpp:
(WebCore::WorkerConsoleClient::profile):
(WebCore::WorkerConsoleClient::profileEnd):
(WebCore::WorkerConsoleClient::takeHeapSnapshot):
(WebCore::WorkerConsoleClient::timeStamp):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::takeHeapSnapshot):
(WebCore::InspectorInstrumentation::consoleTimeStamp):
(WebCore::InspectorInstrumentation::startProfiling):
(WebCore::InspectorInstrumentation::stopProfiling):
Create new `InspectorInstrumentation` hooks for `WorkerGlobalScope` that mimic the behavior of the respective `Page` flow.

* LayoutTests/inspector/worker/resources/worker-timeline.js: Added.
* LayoutTests/inspector/worker/timeline-line-column.html: Added.
* LayoutTests/inspector/worker/timeline-line-column-expected.txt: Added.
Add some basic tests to ensure that `Timeline` works in a `Worker`.

* Source/WebInspectorUI/UserInterface/Test/InspectorProtocol.js:
(InspectorProtocol.getNextMessageId): Added.
(InspectorProtocol.sendCommand):
(InspectorProtocol.awaitCommand):
(InspectorProtocol.awaitMessage):
Expose the message ID for callers to craft their own properly formatted protocol messages.

* Source/WebInspectorUI/UserInterface/Controllers/TimelineManager.js:
(WI.TimelineManager.prototype.initializeTarget):
(WI.TimelineManager.prototype.disable):
(WI.TimelineManager.prototype._updateAutoCaptureInstruments):
* Source/WebInspectorUI/UserInterface/Models/Instrument.js:
(WI.Instrument.startLegacyTimelineAgent):
(WI.Instrument.stopLegacyTimelineAgent):
* Source/WebInspectorUI/UserInterface/Models/ScriptInstrument.js:
(WI.ScriptInstrument.prototype.startInstrumentation):
(WI.ScriptInstrument.prototype.stopInstrumentation):

Drive-by: add missing compatibility comments.
Canonical link: <a href="https://commits.webkit.org/295189@main">https://commits.webkit.org/295189@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f1c75cc6c3a9248d69af7c89d7e55df5fb78755

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104366 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24070 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14485 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/109568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/55034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106406 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24443 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32617 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/109568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/55034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107372 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/19006 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/94186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/109568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/12259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54394 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/12313 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111949 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31523 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/23222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/88279 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/31887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/90430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87944 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22390 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/10611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/25991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31451 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36776 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/31245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/34581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/32805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebCore/workers/WorkerConsoleClient.cpp:
```diff
@@ -89,13 +89,33 @@ void WorkerConsoleClient::timeEnd(JSC::JSGlobalObject* exec, const String& label
-void WorkerConsoleClient::profileEnd(JSC::JSGlobalObject*, const String&) { }
+void WorkerConsoleClient::profile(JSC::JSGlobalObject*, const String& title)
+{
+    // FIXME: Add support for WorkletGlobalScope.
```
I'm not exactly sure of the difference between Worklet and Worker here. Should this just be fixed as part of this change, or does it need a followup bug?

dcrousso replied:
worklets are a more lightweight version of a `Worker` that serve a specific purpose (e.g. `AudioWorklet`)

i dont think we have any Web Inspector support for them right now, but eventually it's something we'd likely want

burg replied:
Okay. Let's get a bugzilla link and put it into these worklet FIXMEs.

burg reviewed (changes requested):
In the changelog, please include more information of how to test this manually.

It seems like this is something we should be able to test in layout tests. Could you port existing tests for this?

burg commented on Source/WebInspectorUI/UserInterface/Controllers/TimelineManager.js:
```diff
@@ -86,6 +86,7 @@ WI.TimelineManager = class TimelineManager extends WI.Object
 
         if (target.hasDomain("Timeline")) {
             // COMPATIBILITY (iOS 13): Timeline.enable did not exist yet.
+            // COMPATIBILITY (iOS X.Y, macOS X.Y): `Timeline.enable` did not exist yet for Worker targets.
```
Won't this fail at `target.hasDomain("Timeline")` in this case?

I believe we'd want to put iOS 18.4 since that's the last versioned protocol file without "worker" targetType for Timeline domain.

dcrousso replied:
oh hmm yeah good catch.  ill change it

we normally only update the compatibility comments once the next protocol version is cut as that's the cleanest way to know for sure when it was first shipped

dcrousso committed:
Web Inspector: support `console.profile` in `Worker`
https://bugs.webkit.org/show_bug.cgi?id=287069

Reviewed by BJ Burg.

* Source/WebCore/workers/WorkerConsoleClient.cpp:
(WebCore::WorkerConsoleClient::profile):
(WebCore::WorkerConsoleClient::profileEnd):
(WebCore::WorkerConsoleClient::takeHeapSnapshot):
(WebCore::WorkerConsoleClient::timeStamp):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::takeHeapSnapshot):
(WebCore::InspectorInstrumentation::consoleTimeStamp):
(WebCore::InspectorInstrumentation::startProfiling):
(WebCore::InspectorInstrumentation::stopProfiling):
Create new `InspectorInstrumentation` hooks for `WorkerGlobalScope` that mimic the behavior of the respective `Page` flow.

* LayoutTests/inspector/worker/resources/worker-timeline.js: Added.
* LayoutTests/inspector/worker/timeline-line-column.html: Added.
* LayoutTests/inspector/worker/timeline-line-column-expected.txt: Added.
Add some basic tests to ensure that `Timeline` works in a `Worker`.

* Source/WebInspectorUI/UserInterface/Test/InspectorProtocol.js:
(InspectorProtocol.getNextMessageId): Added.
(InspectorProtocol.sendCommand):
(InspectorProtocol.awaitCommand):
(InspectorProtocol.awaitMessage):
Expose the message ID for callers to craft their own properly formatted protocol messages.

* Source/WebInspectorUI/UserInterface/Controllers/TimelineManager.js:
(WI.TimelineManager.prototype.initializeTarget):
(WI.TimelineManager.prototype.disable):
(WI.TimelineManager.prototype._updateAutoCaptureInstruments):
* Source/WebInspectorUI/UserInterface/Models/Instrument.js:
(WI.Instrument.startLegacyTimelineAgent):
(WI.Instrument.stopLegacyTimelineAgent):
* Source/WebInspectorUI/UserInterface/Models/ScriptInstrument.js:
(WI.ScriptInstrument.prototype.startInstrumentation):
(WI.ScriptInstrument.prototype.stopInstrumentation):

Drive-by: add missing compatibility comments.
Canonical link: https://commits.webkit.org/295189@main
